File: src/types/patient.ts
import type { WfAppointment } from '#/types/appointment.ts';
import type { DispensingRecord } from '#/types/dispensing.ts';
import type { InrRecord } from '#/types/inr.ts';
import type { WfOutcome } from '#/types/outcome.ts';
import type { WfVisit } from '#/types/visit.ts';

export interface WfPatient {
  id: number;
  hn: string;
  enrolledAt: string;
  enrolledBy?: string;
  status: 'active' | 'inactive' | 'deceased' | 'transferred' | 'discharged';
  indication?: 'AF' | 'DVT' | 'PE' | 'mechanical_valve' | 'other';
  targetInrLow: number;
  targetInrHigh: number;
  notes?: string;
  createdAt: string;
  updatedAt: string;
}

export interface HosxpPatient {
  hn: string;
  pname: string;
  fname: string;
  lname: string;
  birthday: string;
  sex: string;
  addrpart?: string;
  phone?: string;
  informTel?: string;
}

export interface PatientDrugRecord {
  hn: string;
  pname: string;
  fname: string;
  lname: string;
  birthday: string;
  sex: string;
  phone?: string;
  firstDispenseDate: string;
  lastDispenseDate: string;
  totalDispenseVisits: number;
  strengthsReceived: string[];
  isEnrolled: boolean;
  enrollmentStatus?: string;
}

export interface EnrollmentInput {
  hn: string;
  indication: string;
  targetInrLow: number;
  targetInrHigh: number;
  enrolledAt: string;
  enrolledBy: string;
  notes?: string;
}

export interface PatientDetail {
  patient: WfPatient;
  hosxpInfo: HosxpPatient;
  latestInr?: InrRecord;
  currentDoseMgday?: number;
  ttr6months?: number | null;
  nextAppointment?: string;
  inrHistory?: InrRecord[];
  dispensingHistory?: DispensingRecord[];
  visits?: WfVisit[];
  appointments?: WfAppointment[];
  outcomes?: WfOutcome[];
}

export interface ActivePatientSummary {
  patient: WfPatient;
  hosxpInfo: HosxpPatient;
  latestInr?: InrRecord;
  currentDoseMgday?: number;
  ttr6months?: number | null;
  nextAppointment?: string;
  lastVisitDate?: string;
}

File: crates/warfarin-core/src/models/patient.rs
use serde::{Deserialize, Serialize};

use super::{dispensing::DispensingRecord, inr::InrRecord};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WfPatient {
  pub id: i64,
  pub hn: String,
  pub enrolled_at: String,
  pub enrolled_by: Option<String>,
  pub status: String,
  pub indication: Option<String>,
  pub target_inr_low: f64,
  pub target_inr_high: f64,
  pub notes: Option<String>,
  pub created_at: String,
  pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HosxpPatient {
  pub hn: String,
  pub pname: String,
  pub fname: String,
  pub lname: String,
  pub birthday: String,
  pub sex: String,
  pub addrpart: Option<String>,
  pub phone: Option<String>,
  /// Secondary contact number from `HOSxP` `patient.informtel`.
  pub inform_tel: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatientDrugRecord {
  pub hn: String,
  pub pname: String,
  pub fname: String,
  pub lname: String,
  pub birthday: String,
  pub sex: String,
  pub phone: Option<String>,
  pub first_dispense_date: String,
  pub last_dispense_date: String,
  pub total_dispense_visits: usize,
  pub strengths_received: Vec<String>,
  pub is_enrolled: bool,
  pub enrollment_status: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnrollmentInput {
  pub hn: String,
  pub indication: String,
  pub target_inr_low: f64,
  pub target_inr_high: f64,
  pub enrolled_at: String,
  pub enrolled_by: String,
  pub notes: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatientDetail {
  pub patient: WfPatient,
  pub hosxp_info: HosxpPatient,
  pub latest_inr: Option<InrRecord>,
  pub current_dose_mgday: Option<f64>,
  pub ttr6months: Option<f64>,
  pub next_appointment: Option<String>,
  pub inr_history: Vec<InrRecord>,
  pub dispensing_history: Vec<DispensingRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivePatientSummary {
  pub patient: WfPatient,
  pub hosxp_info: HosxpPatient,
  pub latest_inr: Option<InrRecord>,
  pub current_dose_mgday: Option<f64>,
  pub ttr6months: Option<f64>,
  pub next_appointment: Option<String>,
  /// Most recent clinic visit date from `wf_visits`, if any.
  pub last_visit_date: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchFilters {
  pub keyword: Option<String>,
  pub date_from: Option<String>,
  pub date_to: Option<String>,
  pub enrollment_status: Option<String>,
  pub page: u32,
  pub page_size: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchResponse {
  pub items: Vec<PatientDrugRecord>,
  pub total: usize,
}
